crates/ruma-common/src/push/condition/flattened_json.rs

```diff
@@ -0,0 +1,379 @@
+use js_int::Int;
+use serde::{Deserialize, Deserializer, Serialize, Serializer};
+use serde_json::{to_value as to_json_value, value::Value as JsonValue};
+use std::collections::BTreeMap;
+use thiserror::Error;
+use tracing::{instrument, warn};
+
+use crate::serde::Raw;
+
+/// The flattened representation of a JSON object.
+#[derive(Clone, Debug)]
+pub struct FlattenedJson {
+    /// The internal map containing the flattened JSON as a pair path, value.
+    map: BTreeMap<String, FlattenedJsonValue>,
+}
+
+impl FlattenedJson {
+    /// Create a `FlattenedJson` from `Raw`.
+    pub fn from_raw<T>(raw: &Raw<T>) -> Self {
+        let mut s = Self { map: BTreeMap::new() };
+        s.flatten_value(to_json_value(raw).unwrap(), "".into());
+        s
+    }
+
+    /// Flatten and insert the `value` at `path`.
+    #[instrument(skip(self, value))]
+    fn flatten_value(&mut self, value: JsonValue, path: String) {
+        match value {
+            JsonValue::Object(fields) => {
+                for (key, value) in fields {
+                    let key = escape_key(&key);
+                    let path = if path.is_empty() { key } else { format!("{path}.{key}") };
+                    self.flatten_value(value, path);
+                }
+            }
+            value => {
+                if let Ok(v) = value.try_into() {
+                    if self.map.insert(path.clone(), v).is_some() {
+                        warn!("Duplicate path in flattened JSON: {path}");
+                    }
+                }
+            }
+        }
+    }
+
+    /// Value associated with the given `path`.
+    pub fn get(&self, path: &str) -> Option<&FlattenedJsonValue> {
```

Comment:
Can you add `get_str` to shorten all the `get(…).and_then(FlattenedJsonValue::as_str)`s?